File: 백준/Gold/22352. 항체 인식/항체 인식.cc
#include <iostream>
#include <vector>

using namespace std;

int N, M;
vector<vector<int>> SP1, SP2;

bool is_correct_vaccine() {
  for (int i = 0; i < N; i++) {
    for (int j = 0; j < M; j++) {
      if (SP1[i][j] != SP2[i][j]) {
        return false;
      }
    }
  }

  return true;
}

void flood_fill(const int from, const int to, int x, int y) {
  if (x < 0 || x >= N || y < 0 || y >= M || SP1[x][y] != from) {
    return;
  }

  SP1[x][y] = to;
  flood_fill(from, to, x - 1, y);
  flood_fill(from, to, x, y + 1);
  flood_fill(from, to, x + 1, y);
  flood_fill(from, to, x, y - 1);
}

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(nullptr);

  cin >> N >> M;

  SP1.assign(N, vector(M, 0));
  SP2.assign(N, vector(M, 0));

  for (int i = 0; i < N; i++) {
    for (int j = 0; j < M; j++) {
      cin >> SP1[i][j];
    }
  }

  int x, y, from, to;
  for (int i = 0; i < N; i++) {
    for (int j = 0; j < M; j++) {
      cin >> SP2[i][j];

      if (SP1[i][j] != SP2[i][j]) {
        x = i;
        y = j;
        
        from = SP1[i][j];
        to = SP2[i][j];
      }
    }
  }

  flood_fill(from, to, x, y);
  if (is_correct_vaccine()) {
    cout << "YES";
  } else {
    cout << "NO";
  }

  return 0;
}

File: 백준/Gold/22352. 항체 인식/항체 인식.py
input = open(0).readline

N, M = map(int, input().split())
SP1 = [list(map(int, input().split())) for _ in range(N)]
SP2 = [list(map(int, input().split())) for _ in range(N)]

info = next(((SP1[i][j], SP2[i][j], i, j) for i in range(N) for j in range(M) if SP1[i][j] != SP2[i][j]), None)

def flood_fill(before, after, x, y):
    if x < 0 or x >= N or y < 0 or y >= M or SP1[x][y] != before:
        return
    
    SP1[x][y] = after
    
    flood_fill(before, after, x - 1, y)
    flood_fill(before, after, x, y + 1)
    flood_fill(before, after, x + 1, y)
    flood_fill(before, after, x, y - 1)


if info != None:
    flood_fill(*info)

is_correct_vaccine = all(SP1[i][j] == SP2[i][j] for i in range(N) for j in range(M))
    
print('YES' if info == None or is_correct_vaccine else 'NO')